From e6c1646bd964ff4d6bc62fdf1bee48bd784923a9 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentin.casasnovas@vates.tech>
Date: Mon, 23 Mar 2026 10:36:40 +0100
Subject: [PATCH] !kabi Revert: net/sched: flower: fix parsing of ethertype
 following VLAN header

We have no padding/holes in the struct and using shadow variables isn't
straightforward as it is not clear if we should allocate the shadows in
atomic or kernel context and also, the struct contents are used to generate
a hash, so having the shadow outside the main struct would have an effect
on hash.

This is not a critical commit and just a simple bug fix in the vlan code,
so simply revert it.

Reverts: f0c8a97c6019 ("net/sched: flower: fix parsing of ethertype following VLAN header")
Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 include/net/flow_dissector.h |  2 --
 net/core/flow_dissector.c    |  1 -
 net/sched/cls_flower.c       | 18 +++++-------------
 3 files changed, 5 insertions(+), 16 deletions(-)

diff --git a/include/net/flow_dissector.h b/include/net/flow_dissector.h
index 6a0165bdc25f..844bad1ca3a8 100644
--- a/include/net/flow_dissector.h
+++ b/include/net/flow_dissector.h
@@ -50,8 +50,6 @@ struct flow_dissector_key_vlan {
 	u16	vlan_id:12,
 		vlan_priority:3;
 	__be16	vlan_tpid;
-	__be16	vlan_eth_type;
-	u16	padding;
 };
 
 struct flow_dissector_key_mpls {
diff --git a/net/core/flow_dissector.c b/net/core/flow_dissector.c
index da860a680256..949694c70cbc 100644
--- a/net/core/flow_dissector.c
+++ b/net/core/flow_dissector.c
@@ -827,7 +827,6 @@ bool __skb_flow_dissect(const struct sk_buff *skb,
 					 VLAN_PRIO_MASK) >> VLAN_PRIO_SHIFT;
 			}
 			key_vlan->vlan_tpid = saved_vlan_tpid;
-			key_vlan->vlan_eth_type = proto;
 		}
 
 		fdret = FLOW_DISSECT_RET_PROTO_AGAIN;
diff --git a/net/sched/cls_flower.c b/net/sched/cls_flower.c
index 7ffa28a98d74..d0ed0ad203d1 100644
--- a/net/sched/cls_flower.c
+++ b/net/sched/cls_flower.c
@@ -554,7 +554,6 @@ static int fl_set_key_mpls(struct nlattr **tb,
 static void fl_set_key_vlan(struct nlattr **tb,
 			    __be16 ethertype,
 			    int vlan_id_key, int vlan_prio_key,
-			    int vlan_next_eth_type_key,
 			    struct flow_dissector_key_vlan *key_val,
 			    struct flow_dissector_key_vlan *key_mask)
 {
@@ -573,11 +572,6 @@ static void fl_set_key_vlan(struct nlattr **tb,
 	}
 	key_val->vlan_tpid = ethertype;
 	key_mask->vlan_tpid = cpu_to_be16(~0);
-	if (tb[vlan_next_eth_type_key]) {
-		key_val->vlan_eth_type =
-			nla_get_be16(tb[vlan_next_eth_type_key]);
-		key_mask->vlan_eth_type = cpu_to_be16(~0);
-	}
 }
 
 static void fl_set_key_flag(u32 flower_key, u32 flower_mask,
@@ -810,9 +804,8 @@ static int fl_set_key(struct net *net, struct nlattr **tb,
 
 		if (eth_type_vlan(ethertype)) {
 			fl_set_key_vlan(tb, ethertype, TCA_FLOWER_KEY_VLAN_ID,
-					TCA_FLOWER_KEY_VLAN_PRIO,
-					TCA_FLOWER_KEY_VLAN_ETH_TYPE,
-					&key->vlan, &mask->vlan);
+					TCA_FLOWER_KEY_VLAN_PRIO, &key->vlan,
+					&mask->vlan);
 
 			if (tb[TCA_FLOWER_KEY_VLAN_ETH_TYPE]) {
 				ethertype = nla_get_be16(tb[TCA_FLOWER_KEY_VLAN_ETH_TYPE]);
@@ -820,7 +813,6 @@ static int fl_set_key(struct net *net, struct nlattr **tb,
 					fl_set_key_vlan(tb, ethertype,
 							TCA_FLOWER_KEY_CVLAN_ID,
 							TCA_FLOWER_KEY_CVLAN_PRIO,
-							TCA_FLOWER_KEY_CVLAN_ETH_TYPE,
 							&key->cvlan, &mask->cvlan);
 					fl_set_key_val(tb, &key->basic.n_proto,
 						       TCA_FLOWER_KEY_CVLAN_ETH_TYPE,
@@ -1728,13 +1720,13 @@ static int fl_dump_key(struct sk_buff *skb, struct net *net,
 		goto nla_put_failure;
 
 	if (mask->basic.n_proto) {
-		if (mask->cvlan.vlan_eth_type) {
+		if (mask->cvlan.vlan_tpid) {
 			if (nla_put_be16(skb, TCA_FLOWER_KEY_CVLAN_ETH_TYPE,
 					 key->basic.n_proto))
 				goto nla_put_failure;
-		} else if (mask->vlan.vlan_eth_type) {
+		} else if (mask->vlan.vlan_tpid) {
 			if (nla_put_be16(skb, TCA_FLOWER_KEY_VLAN_ETH_TYPE,
-					 key->vlan.vlan_eth_type))
+					 key->basic.n_proto))
 				goto nla_put_failure;
 		}
 	}
-- 
2.43.0

